File: src/utils/textSearch.ts
/**
 * textSearch.ts
 * Thuật toán tìm kiếm linh động chuẩn tiếng Việt phong cách Listary:
 * - Bỏ dấu tiếng Việt (kể cả đ/Đ).
 * - Khớp cả cụm liền nhau.
 * - Khớp từng từ không cần đúng thứ tự.
 * - Chấp nhận sai chính tả / khớp mờ (fuzzy).
 */

const EXTRA_MAP: Record<string, string> = {
  'đ': 'd',
  'Đ': 'D',
};

export function stripDiacritics(text: string): string {
  if (!text) return '';
  let str = text;
  for (const [k, v] of Object.entries(EXTRA_MAP)) {
    str = str.replaceAll(k, v);
  }
  return str
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '')
    .normalize('NFC');
}

export function normalize(text: string): string {
  if (!text) return '';
  const clean = stripDiacritics(text).toLowerCase();
  return clean.replace(/\s+/g, ' ').trim();
}

export function tokenize(normalizedText: string): string[] {
  const matches = normalizedText.match(/[^\W_]+/gu);
  return matches ? Array.from(matches) : [];
}

function stringSimilarity(s1: string, s2: string): number {
  if (s1 === s2) return 1.0;
  if (s1.includes(s2) || s2.includes(s1)) return 0.92;
  const longer = s1.length > s2.length ? s1 : s2;
  const shorter = s1.length > s2.length ? s2 : s1;
  const longerLength = longer.length;
  if (longerLength === 0) return 1.0;

  // Levenshtein distance approximation for short tokens
  const costs: number[] = [];
  for (let i = 0; i <= s1.length; i++) {
    let lastValue = i;
    for (let j = 0; j <= s2.length; j++) {
      if (i === 0) {
        costs[j] = j;
      } else if (j > 0) {
        let newValue = costs[j - 1];
        if (s1.charAt(i - 1) !== s2.charAt(j - 1)) {
          newValue = Math.min(Math.min(newValue, lastValue), costs[j]) + 1;
        }
        costs[j - 1] = lastValue;
        lastValue = newValue;
      }
    }
    if (i > 0) costs[s2.length] = lastValue;
  }
  return (longerLength - costs[s2.length]) / longerLength;
}

function bestTokenRatio(token: string, targetTokens: string[]): number {
  let best = 0.0;
  for (const t of targetTokens) {
    if (token === t) return 1.0;
    if (best < 0.92) {
      if (token.includes(t) || t.includes(token)) {
        best = Math.max(best, 0.92);
        continue;
      }
      const ratio = stringSimilarity(token, t);
      if (ratio > best) {
        best = ratio;
      }
    }
  }
  return best;
}

export function matchScore(query: string, ...fields: string[]): number {
  const queryNorm = normalize(query);
  if (!queryNorm) return 1.0;

  const combinedNorm = normalize(fields.filter(Boolean).join(' '));
  if (!combinedNorm) return 0.0;

  let score = 0.0;

  // Khớp cả cụm liền nhau
  if (combinedNorm.includes(queryNorm)) {
    score += 100.0;
    if (combinedNorm.startsWith(queryNorm)) {
      score += 20.0;
    }
  }

  const qTokens = tokenize(queryNorm);
  if (qTokens.length === 0) return score;

  const targetTokens = tokenize(combinedNorm);
  if (targetTokens.length === 0) return score;

  const perTokenScores = qTokens.map((tok) => bestTokenRatio(tok, targetTokens));
  const matched = perTokenScores.filter((r) => r >= 0.7);

  if (matched.length < qTokens.length) {
    return score;
  }

  score += matched.reduce((a, b) => a + b, 0) * 10.0;
  return score;
}

File: text_search.py
# -*- coding: utf-8 -*-
"""
text_search.py
================
Tiện ích tìm kiếm "linh động" kiểu Listary, có hỗ trợ tiếng Việt:

  - Không phân biệt hoa/thường.
  - Gõ KHÔNG DẤU vẫn tìm ra nội dung CÓ DẤU (vd: "mat bang" -> "Mặt bằng").
  - Không cần gõ liền một cụm đúng thứ tự: gõ nhiều từ cách nhau bằng
    khoảng trắng, mỗi từ chỉ cần khớp (đúng, khớp một phần, hoặc gần đúng)
    với MỘT từ bất kỳ trong tiêu đề / mục cha / tên file, không cần đúng
    thứ tự xuất hiện.
  - Chấp nhận sai chính tả nhẹ (khớp mờ - fuzzy) nhờ so khớp tỉ lệ tương
    đồng giữa các từ, hữu ích khi gõ nhanh hoặc nhớ không chính xác tên.

Không phụ thuộc thư viện ngoài (chỉ dùng các module chuẩn của Python) để dễ
tái sử dụng và kiểm thử độc lập.
"""

from __future__ import annotations

import difflib
import re
import unicodedata
from functools import lru_cache
from typing import List, Sequence, Tuple

_EXTRA_MAP = str.maketrans({"đ": "d", "Đ": "D"})
_WORD_RE = re.compile(r"[^\W_]+", re.UNICODE)
_FUZZY_THRESHOLD = 0.72
_GOOD_ENOUGH = 0.92


def _is_nan(value) -> bool:
    return isinstance(value, float) and value != value


def _clean(value) -> str:
    if value is None or _is_nan(value):
        return ""
    return str(value)


@lru_cache(maxsize=20000)
def strip_diacritics(text: str) -> str:
    """Bỏ dấu tiếng Việt: 'Mặt bằng tổng thể' -> 'Mat bang tong the'."""
    if not text:
        return ""
    text = text.translate(_EXTRA_MAP)
    decomposed = unicodedata.normalize("NFD", text)
    without_marks = "".join(ch for ch in decomposed if unicodedata.category(ch) != "Mn")
    return unicodedata.normalize("NFC", without_marks)


@lru_cache(maxsize=20000)
def normalize(text: str) -> str:
    """Chuẩn hoá để so khớp: bỏ dấu, chữ thường, gộp khoảng trắng thừa."""
    if not text:
        return ""
    text = strip_diacritics(text).lower()
    return re.sub(r"\s+", " ", text).strip()


@lru_cache(maxsize=20000)
def tokenize(normalized_text: str) -> Tuple[str, ...]:
    return tuple(_WORD_RE.findall(normalized_text))


def _best_token_ratio(token: str, target_tokens: Sequence[str]) -> float:
    best = 0.0
    for t in target_tokens:
        if token == t:
            return 1.0
        if best < _GOOD_ENOUGH:
            if token in t or t in token:
                best = max(best, _GOOD_ENOUGH)
                continue
            ratio = difflib.SequenceMatcher(None, token, t).ratio()
            if ratio > best:
                best = ratio
    return best


def match_score(query: str, *fields: str) -> float:
    """
    Điểm khớp giữa từ khoá `query` và các trường văn bản `fields`.
    0.0  = không khớp (sẽ bị lọc bỏ khỏi kết quả)
    > 0  = có khớp, càng cao càng liên quan (xếp hạng Listary).
    """
    query_norm = normalize(_clean(query))
    if not query_norm:
        return 1.0

    combined_norm = normalize(" \u241F ".join(_clean(f) for f in fields if _clean(f)))
    if not combined_norm:
        return 0.0

    score = 0.0
    if query_norm in combined_norm:
        score += 100.0
        if combined_norm.startswith(query_norm):
            score += 20.0

    q_tokens = tokenize(query_norm)
    if not q_tokens:
        return score

    target_tokens = tokenize(combined_norm)
    if not target_tokens:
        return score

    per_token_scores = [_best_token_ratio(tok, target_tokens) for tok in q_tokens]
    matched = [r for r in per_token_scores if r >= _FUZZY_THRESHOLD]

    if len(matched) < len(q_tokens):
        return score

    score += sum(matched) * 10.0
    return score


def rank(query: str, rows: Sequence, field_getter) -> List[Tuple[int, float]]:
    scored = []
    for i, row in enumerate(rows):
        s = match_score(query, *field_getter(row))
        if s > 0:
            scored.append((i, s))
    scored.sort(key=lambda pair: pair[1], reverse=True)
    return scored
